subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/BasicJavaCompilerIntegrationSpec.groovy

```diff
@@ -252,91 +252,171 @@ class Main {
+                compileOnly project(":processor")
+            }
+        """
+        settingsFile << "include 'annotation', 'processor'"
+        writeAnnotationProcessorProjects("annotation", "processor", "SimpleAnnotation", "Generated")
+
+        file("src/main/java/Java.java") << "@com.test.SimpleAnnotation public class Java {}"
+
+        then:
+        succeeds("compileJava")
+        file('build/classes/main/Java$$Generated.java').exists()
+    }
+
+    @LeaksFileHandles("holds processor.jar open for in process compiler")
+    def "can use annotation processor in processorpath"() {
+        when:
+        buildFile << """
+            apply plugin: "java"
+            configurations {
+                annotationProcessors
+            }
+            dependencies {
+                compileOnly project(":annotation")
+                annotationProcessors project(":processor")
+            }
+            compileJava {
+                options.processorpath = project.configurations.annotationProcessors
+            }
+        """
+        settingsFile << "include 'annotation', 'processor'"
+        writeAnnotationProcessorProjects("annotation", "processor", "SimpleAnnotation", "Generated")
+
+        file("src/main/java/Java.java") << "@com.test.SimpleAnnotation public class Java {}"
+
+        then:
+        succeeds("compileJava")
+        file('build/classes/main/Java$$Generated.java').exists()
+    }
+
+    @LeaksFileHandles("holds processor.jar open for in process compiler")
+    def "processorpath has precedence over classpath"() {
+        when:
+        buildFile << """
+            apply plugin: "java"
+            configurations {
+                annotationProcessors
+            }
+            dependencies {
+                compileOnly project(":annotation1")
+                compileOnly project(":processor2")
+                annotationProcessors project(":processor1")
+            }
+            compileJava {
+                options.processorpath = project.configurations.annotationProcessors
+                options.compilerArgs += [ '-Xlint:-processing' ] // SimpleAnnotation2 won't be claimed
+            }
+        """
+        settingsFile << "include 'annotation1', 'processor1', 'annotation2', 'processor2'"
+        writeAnnotationProcessorProjects("annotation1", "processor1", "SimpleAnnotation1", "Generated1")
+        writeAnnotationProcessorProjects("annotation2", "processor2", "SimpleAnnotation2", "Generated2")
+
+        file("src/main/java/Java.java") << "@com.test.SimpleAnnotation1 @com.test.SimpleAnnotation2 public class Java {}"
+
+        then:
+        succeeds("compileJava")
+        file('build/classes/main/Java$$Generated1.java').exists()
+        !file('build/classes/main/Java$$Generated2.java').exists()
+    }
+
+    def writeAnnotationProcessorProjects(String annotationProject, String processorProject, String annotationName, String generatedClassSuffix) {
+        file(annotationProject).create {
```

Comment:
To keep the test classes small and readable you should prefer to use the `TestResources` JUnit Rule to organize test project code. Check the JUnitIntegrationTest as an example how to use this Rule.